[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

simonwallis2 opened the pull request:
Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes;
Add some missing LLVM AArch64 OperandTypes.

Llvm-exegesis can now handle many more AArch64 instructions.

AArch64 load/store instructions are not yet supported by llvm-exegesis, until https://github.com/llvm/llvm-project/pull/144895 lands.

Commit:
[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes

Add some missing LLVM AArch64 OperandTypes

Llvm-exegesis can now handle many more instructions.

Load/store instructions not yet supported by llvm-exegesis,
until https://github.com/llvm/llvm-project/pull/144895 lands.

Change-Id: I945079ae290395a207e4cc4619d177a67f0d5a47

Commit:
[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes

Add some missing LLVM AArch64 OperandTypes

Llvm-exegesis can now handle many more instructions.

Load/store instructions not yet supported by llvm-exegesis,
until https://github.com/llvm/llvm-project/pull/144895 lands.

Change-Id: I73d22942943ee5882e89852f538cf1805645def5

Commit:
[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes

Add some missing LLVM AArch64 OperandTypes

Llvm-exegesis can now handle many more instructions.

Load/store instructions not yet supported by llvm-exegesis,
until https://github.com/llvm/llvm-project/pull/144895 lands.

Change-Id: I344efdcfa66b05321125f21c20f7556da583e7ee

simonwallis2 commented:
I see little benefit in adding new tests for this commit.
There is reasonable coverage already from llvm-project/llvm/test/tools/llvm-exegesis/AArch64/all-opcodes.test

boomanaiden154 reviewed (commented):
Regarding test coverage, a couple thoughts:
1. It would be good if we at least assert something in a test, even if that is just the number of snippets successfully generated with a specific random seed so that there isn't back slide.
2. You don't gain much beyond a simple assertion. Asserting the machine code generated might not be a terrible idea, but I would expect the machine verifier to pick up most of the big issues.

simonwallis2 commented on llvm/tools/llvm-exegesis/lib/AArch64/Target.cpp:
```diff
@@ -80,6 +87,42 @@ loadFPCRImmediate(MCRegister Reg, unsigned RegBitWidth, const APInt &Value) {
+static std::vector<MCInst> loadWSeqPairImmediate(MCRegister Reg,
+                                                 unsigned RegBitWidth,
+                                                 const APInt &Value) {
+  MCRegister EvenReg = (Reg - AArch64::W0_W1) * 2 + AArch64::W0 + 0;
```
There is prior art elsewhere in this file, in the handling of FPR8RegClass.
Yes I believe this use of MCRegister arithmetic is standard. It relies on members of the reg classes being contiguous. This is OK because they are contiguous.

simonwallis2 commented on llvm/tools/llvm-exegesis/lib/AArch64/Target.cpp:
```diff
@@ -106,6 +149,110 @@ static MCInst loadFPImmediate(MCRegister Reg, unsigned RegBitWidth,
 
+// Generates instructions to load an immediate value into a DD Reg
+static std::vector<MCInst>
+loadDDRegImmediate(MCRegister Reg, unsigned RegBitWidth, const APInt &Value) {
```
Good idea. Will do.

Commit:
[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes

Add some missing LLVM AArch64 OperandTypes

Llvm-exegesis can now handle many more instructions.

Load/store instructions not yet supported by llvm-exegesis,
until https://github.com/llvm/llvm-project/pull/144895 lands.

Change-Id: I076b2acbb507aa26042d30e821d5d22a8272e369

Commit:
Merge branch 'main' into exegesis-operand-types

Commit:
[llvm-exegesis] Add AArch64 operand initializers, SetRegTo

Make llvm-exegesis more usable on AArch64 by doing the following:

Add some missing exegesis handling of register classes

Add some missing LLVM AArch64 OperandTypes

Llvm-exegesis can now handle many more instructions.

Load/store instructions not yet supported by llvm-exegesis,
until https://github.com/llvm/llvm-project/pull/144895 lands.

Change-Id: Iaae2e4e365c2e5a79ddfb7f4b91ed7e098af46aa

boomanaiden154 reviewed (approved):
Exegesis changes LGTM. Please get someone who knows the ARM side of things well to approve that side of things before landing.

simonwallis2 commented on llvm/tools/llvm-exegesis/lib/AArch64/Target.cpp:
```diff
@@ -224,6 +395,12 @@ Error ExegesisAArch64Target::randomizeTargetMCOperand(
   case llvm::AArch64::OPERAND_IMPLICIT_IMM_0:
     AssignedValue = MCOperand::createImm(0);
     return Error::success();
+  case llvm::AArch64::OPERAND_SHIFTED_REGISTER:
```
Added a TODO comment.

c-rhodes committed:
add setReg_init_check tests

sjoerdmeijer reviewed (approved):
Thanks, LGTM too.

alazarev commented:
Looks like it broke buildbot:
https://lab.llvm.org/buildbot/#/builders/24/builds/16298

c-rhodes commented:
> Looks like it broke buildbot: https://lab.llvm.org/buildbot/#/builders/24/builds/16298

thanks for heads up, might take me a little while to reproduce so I'm just going to go ahead and revert.

c-rhodes commented:
> > Looks like it broke buildbot: https://lab.llvm.org/buildbot/#/builders/24/builds/16298
> 
> thanks for heads up, might take me a little while to reproduce so I'm just going to go ahead and revert.

Managed to reproduce it locally. It appears the snippet varies between normal and ASAN build when using `--random-generator-seed=<seed>`:

ASAN:
```
➜  test /home/culrho01/scratch_dir/llvm_build_asan_ubsan/bin/llvm-exegesis -mtriple=aarch64 -mcpu=neoverse-v2 -mode=latency --dump-object-to-disk=/dev/null --opcode-name=CASPW --benchmark-phase=assemble-measured-code --random-generator-seed=2
Check generated assembly with: /usr/bin/objdump -d /dev/null
---
mode:            latency
key:
  instructions:
    - 'CASPW W16_W17 W16_W17 W26_W27 X7'
  config:          ''
  register_initial_values:
    - 'W16_W17=0x0'
    - 'W26_W27=0x0'
    - 'X7=0x0'
cpu_name:        neoverse-v2
llvm_triple:     aarch64
min_instructions: 10000
measurements:    []
error:           actual measurements skipped.
info:            Repeating a single implicitly serial instruction
assembled_snippet: FB6BBFA910008052110080521A0080521B008052070080D2FA7C3008FA7C3008FA7C3008FA7C3008FB6BC1A8C0035FD6
...
```
Standard:
```
➜  test /home/culrho01/llvm-project/build/bin/llvm-exegesis -mtriple=aarch64 -mcpu=neoverse-v2 -mode=latency --dump-object-to-disk=/dev/null --opcode-name=CASPW --benchmark-phase=assemble-measured-code --random-generator-seed=2
Check generated assembly with: /usr/bin/objdump -d /dev/null
---
mode:            latency
key:
  instructions:
    - 'CASPW W12_W13 W12_W13 W4_W5 LR'
  config:          ''
  register_initial_values:
    - 'W12_W13=0x0'
    - 'W4_W5=0x0'
    - 'LR=0x0'
cpu_name:        neoverse-v2
llvm_triple:     aarch64
min_instructions: 10000
measurements:    []
error:           actual measurements skipped.
info:            Repeating a single implicitly serial instruction
assembled_snippet: FE0F1FF80C0080520D00805204008052050080521E0080D2C47F2C08C47F2C08C47F2C08C47F2C08FE0741F8C0035FD6
...
```
